openssl/src/pkcs7.rs

```diff
@@ -146,6 +183,82 @@ impl Pkcs7 {
             .map(Pkcs7)
         }
     }
+
+    /// Set signed attributes in a PKCS#7 structure.
+    ///
+    /// `attributes` is a stack of the attributes to be added.
+    ///
+    #[corresponds(PKCS7_set_signed_attributes)]
+    pub fn set_signed_attributes(
+        signer_info: &Pkcs7SignerInfoRef,
+        attributes: Stack<X509Attribute>,
+    ) -> Result<(), ErrorStack> {
+        unsafe {
+            cvt(ffi::PKCS7_set_signed_attributes(
+                signer_info.as_ptr(),
+                attributes.as_ptr(),
+            ))?;
+            mem::forget(attributes);
+            Ok(())
+        }
+    }
+
+    /// Add a signed attribute to a PKCS7_SIGNER_INFO structure
+    ///
+    /// `nid` is the Nid of the attribute, `atrtype` is the attribute's type (an ASN.1 tag
+    /// value) and `value` is the ASN1 object to be added.
+    ///
+    //  Note: `value` is immutable, but the OpenSSL function takes a (non-const) void pointer. Thus,
+    //  we have to cast to mutable in the unsafe block.
+    //  OpenSSL takes ownership of `value`.
+    //
+    #[corresponds(PKCS7_add_signed_attribute)]
+    pub fn add_signed_attribute(
+        signer_info: &Pkcs7SignerInfoRef,
+        nid: Nid,
+        atrtype: Asn1Type,
+        value: Asn1Object,
+    ) -> Result<(), ErrorStack> {
+        unsafe {
+            cvt(ffi::PKCS7_add_signed_attribute(
+                signer_info.as_ptr(),
+                nid.as_raw(),
+                atrtype.as_raw(),
+                value.as_ptr() as *mut c_void,
+            ))?;
+            mem::forget(value);
+            Ok(())
+        }
+    }
+
+    /// Get the certificates stored to a PKCS#7 structure
+    ///
+    // Unfortunately, there is no corresponding function in openssl. Thus, we have to enter
+    // openssl's internal PKCS7 struct. This is also, what openssl does, when e.g.
+    // `openssl pkcs7 -in pkcs7.pem -print-certs` is called.
+    //
+    pub fn certificates(&self) -> Result<Stack<X509>, ErrorStack> {
```

Comment:
I would prefer to not build nontrivial APIs that don't have an equivalent in OpenSSL itself. It might be preferable to expose the various components of the signer info and allow this logic to be written externally.

Reply:
This case is a little delicate. As I mentioned in my first comment, OpenSSL api does not provide access to the certificates `STACK` of a `PKCS7` of type `PKCS7_SIGNED` or `PKCS7_SIGNEDANDENVELOPED`. But I don't think, that's missing on purpose. OpenSSL itself accesses this directly and SSCEP does that, too. Because they need it and there's no other way. I chose to return a (new) `Stack<X509>` (instead of a `&StackRef<X509>`), because I wanted to avoid a dangling reference. But probably, there is a better way?